rclcpp/src/rclcpp/node_interfaces/node_graph.cpp

```diff
@@ -369,3 +369,103 @@ NodeGraph::count_graph_users()
 {
   return graph_users_count_.load();
 }
+
+static
+rclcpp::TopicInfo
+convert_to_topic_info(const rmw_topic_info_t * info)
+{
+  rclcpp::TopicInfo topic_info;
+  topic_info.node_name = info->node_name;
+  topic_info.node_namespace = info->node_namespace;
+  topic_info.topic_type = info->topic_type;
+  memcpy(topic_info.gid, info->gid, RMW_GID_STORAGE_SIZE);
+  memcpy(&topic_info.qos_profile, &info->qos_profile, sizeof(rmw_qos_profile_t));
+  return topic_info;
+}
+
+static
+std::vector<rclcpp::TopicInfo>
+convert_to_topic_info_list(const rmw_topic_info_array_t * info_array)
+{
+  std::vector<rclcpp::TopicInfo> topic_info_list;
+  for (size_t i = 0; i < info_array->count; ++i) {
+    rmw_topic_info_t topic_info = info_array->info_array[i];
+    auto topic_info_item = convert_to_topic_info(&topic_info);
+    topic_info_list.push_back(topic_info_item);
+  }
+  return topic_info_list;
+}
+
+std::vector<rclcpp::TopicInfo>
+NodeGraph::get_info_by_topic(
+  const std::string & topic_name,
+  bool no_mangle,
+  const std::string & type,
+  rcl_get_info_by_topic_func_t rcl_get_info_by_topic) const
+{
+  std::vector<rclcpp::TopicInfo> topic_info_list;
+
+  auto rcl_node_handle = node_base_->get_rcl_node_handle();
+  auto fqdn = rclcpp::expand_topic_or_service_name(
+    topic_name,
+    rcl_node_get_name(rcl_node_handle),
+    rcl_node_get_namespace(rcl_node_handle),
+    false);    // false = not a service
```

Comment:
> What specifically did you mean should be done in the rcl layer? This whole function?

Yes.

> Is rclcpp::expand_topic_or_service_name() getting the fully qualified name of topic_name, and there's an equivalent in rcl?

Is getting a name and converting it to a "fully qualified name" (actually, it's just expanding, remappings are missing).
In `rcl`, there is a function that does the `expand`, and another that does the remapping. That combination is repeated in many place in the codebase. I think it would be good to combine the whole `expand and remap` logic in one function (a `is_service` flag will be needed).

----

Most `rcl` functions, like `create_publisher`, accept a topic name that's then expanded and remapped.
The only exception I know is `rcl_count_publishers`, which requires a fully qualified name.
The `rclcpp` counterpart doesn't requires a fully qualified name, but does the expansion (and, AFAIS the remapping part is missing). I wonder if both the expansion and remapping should be done in `rcl`, which would improve code reusage.

Reply:
> I wonder if both the expansion and remapping should be done in `rcl`, which would improve code reusage.

Yeah, my first thought was that it's surprising that there's a `expand_topic_or_service_name()`
in `rclcpp`. I thought it was going to be at most just a thin wrapper around something in `rcl`.

Reply:
I agree that "rclcpp::expand_topic_or_service_name()" should not at rclcpp layer.  
rclcpp::expand_topic_or_service_name() not only expand the name, but to check if the name is valid and then throw exception of rclcpp.  
If not calling "rclcpp::expand_topic_or_service_name()" at rclcpp layer, I think rcl should check name at each interfaces.  
and then rclcpp check the returned value from rcl to throw rclcpp itself exception.  
Maybe need a new issue to handle this.

Reply:
So to answer @ivanpauno's question

> I wonder if this should be done in rcl layer.

Yeah, I think so, but it's outside the scope of this PR.

Reply:
Sounds reasonable, the changes are broader that just this new functions.

Reply:
Is this resolved?

Reply:
Following my comment in the conversation about the line just above, I think we need to decide if it is needed or not, and if it is needed then we should probably do it in rcl if we can, but I agree we could do it in a follow up. However, I would like to have a decision on what we should do before continuing here.

Reply:
I think there's a couple of things going on here.

`rclcpp::expand_topic_or_service_name()` does topic name expansion and validation, and does not do remapping.

`rclpy.expand_topic_name.expand_topic_name()` does topic name expansion, does not do validation (the caller is supposed to do it manually), and also does not do remapping.

(My understanding of remapping here is that the returned/output topic name is the remapped version of the input topic name.)

The follow-up should be to implement in `rcl` a function like `rcl_get_expanded_and_remapped_topic_name()` (and it also does validation). This function supposedly should be agnostic to the `no_mangle` parameter as well.